tools/launch.py

```diff
@@ -424,6 +432,68 @@ def wrap_cmd_with_extra_envvars(cmd: str, env_vars: list) -> str:
     env_vars = " ".join(env_vars)
     return wrap_cmd_with_local_envvars(cmd, env_vars)
 
+
+g_monitor_file = None
+g_group_id = 0
+
+def has_alive_servers(args):
+    """Check whether there exist alive servers and get available group ID if exist.
+
+    For each group of long live servers, a monitor file named
+    'dgl_dist_monitor_{args.server_name}' is created under '/tmp/' directory.
+    We check the existence of this monitor file to determine whether to
+    launch new servers or utilize the existing alive ones. If there
+    exist alive servers, we obtain availale group ID from the monitor
+    file which could be used in current client groups.
+
+    """
+    if args.server_name is None:
+        return False
+    global g_monitor_file
+    global g_group_id
+    monitor_file = '/tmp/dgl_dist_monitor_' + args.server_name
+    from filelock import FileLock
+    lock = FileLock(monitor_file + '.lock')
+    with lock:
+        next_group_id = None
+        ret = os.path.exists(monitor_file)
+        if ret:
+            print("Monitor file for alive servers already exist: {}.".format(monitor_file))
+            lines = [line.rstrip('\n') for line in open(monitor_file)]
+            g_group_id = int(lines[0])
+            next_group_id = g_group_id + 1
+        if not ret and args.keep_alive:
+            next_group_id = 1
+            print("Monitor file for alive servers is created: {}.".format(monitor_file))
+            g_monitor_file = monitor_file
+        if next_group_id is not None:
+            with open(monitor_file, 'w') as f:
+                f.write(str(next_group_id))
+    return ret
+
+
+def clean_alive_servers():
+    """Remove keep alive related files"""
+    global g_monitor_file
+    try:
+        if g_monitor_file is not None:
+            os.remove(g_monitor_file)
+            os.remove(g_monitor_file + '.lock')
+            print("Monitor file for alive servers is removed: {}.".format(g_monitor_file))
+    except:
+        print("Failed to delete monitor file for alive servers: {}.".format(g_monitor_file))
+
+def get_available_port(ip):
+    """Get available port with specified ip."""
```

Comment:
This post may be useful for finding an available port: https://stackoverflow.com/questions/1365265/on-localhost-how-do-i-pick-a-free-port-number

Reply:
the post you mentioned is helpful for getting available port on current host while we need available port for a specified host. we may have multiple hosts.